src/hotbaud/experimental/efdlock.py

```diff
@@ -0,0 +1,129 @@
+'''
+IPC/thread/task Lock
+
+Requires EFD_SEMAPHORE on eventfd open
+
+TODO: docstring
+
+TODO: considerations of using EFD_NONBLOCK on open: will force sync readers to do polling
+
+'''
+import os
+
+from typing import Generator, Type
+from contextlib import contextmanager
+
+import trio
+from hotbaud.eventfd import EFD_NONBLOCK, EFD_SEMAPHORE, open_eventfd, read_eventfd, write_eventfd
+
+
+class LockCommon:
+    '''
+    Common between sync & async Lock impls
+
+    '''
+    def __init__(self, fd: int):
+        self._fd = fd
+        self._locked = False
+
+    @property
+    def fd(self) -> int:
+        return self._fd
+
+    @property
+    def locked(self) -> bool:
+        return self._locked
+
+    def _ensure_locked(self) -> None:
+        '''
+        Raise if the lock is not held.
+
+        '''
+        if not self._locked:
+            raise RuntimeError('Tried to release un-acquired lock')
+
+    def release(self) -> None:
+        self._ensure_locked()
+        self._locked = False
+        write_eventfd(self._fd, 1)
+
+
+class Lock(LockCommon):
+    def acquire(self) -> None:
```

Comment:
I want to say the same thing i did in `event.py`, maybe `.acquire_sync()` is cool here and then we can minimize the type hierarchy?

I do know (bc of a recent attempt 😂:joy:) that you can't use `contextlib.ContextManagerDecorator` and the async version in the type hierarchy but pretty sure `__[a]enter__()` can be on the same type?  So in this case 2 types with the `AsyncLock` having the async dunder meth defined as extra?